src/filters/filters.ts

```diff
@@ -0,0 +1,263 @@
+import { Datum } from "plotly.js";
+import { linearRegressionLine, linearRegression } from "simple-statistics";
+import {
+  parseTimeDuration,
+  TimeDurationStr,
+  timeUnits,
+} from "../duration/duration";
+import { StatisticValue } from "../recorder-types";
+import { HassEntity, YValue } from "../types";
+
+const castFloat = (y: any) => parseFloat(y);
+const myEval = typeof window != "undefined" ? window.eval : global.eval;
+
+type FilterData = {
+  xs: Date[];
+  ys: YValue[];
+  states: HassEntity[];
+  statistics: StatisticValue[];
+  meta: HassEntity["attributes"];
+  vars: Record<any, any>;
+};
+export type FilterFn = (p: FilterData) => Partial<FilterData>;
+
+const mapNumbers = (ys: YValue[], fn: (y: number, i: number) => number) =>
+  ys.map((y, i) => {
+    const n = castFloat(y);
+    if (Number.isNaN(n)) return y;
+    return fn(n, i);
+  });
+
+/**
+ * Removes from all params the indexes for which ys is not numeric, and parses ys to numbers.
+ * WARNING: when used inside a filter, it is important to return all arrays. Otherwise the lengths
+ * between say ys and states won't be consistent
+ */
+const force_numeric: (p: FilterData) => { ys: number[] } & FilterData = ({
+  xs,
+  ys: ys2,
+  states,
+  statistics,
+  ...rest
+}) => {
+  const ys = ys2.map((y) => castFloat(y));
+  const mask = ys.map((y) => !isNaN(y));
+  return {
```

Comment:
I like the implementation, but I don't agree with the choice of filtering non-numbers.
If I understood it correctly this _removes_ null values, practically making them disappear. This can cause a false representation of reality.

ESPHome struggled with that some time ago, see here:
- https://github.com/esphome/issues/issues/3338
  - https://github.com/esphome/issues/issues/3338#issuecomment-1148956927
- https://github.com/esphome/esphome/pull/3610

Reply:
Oh, yes. I struggled a lot to find a compromise between code clarity and keeping the nulls intact. Force_numeric is only used by the averaging filters, whose whole point is to remove anomalies and smooth out traces. All the other ones handle the null by skipping them.
I also like how the implementation looks like, way cleaner than rebuilding the arrays from scratch in a for loop.

Reply:
The rolling window averaging is the one that is most incompatible with nulls, as even the x scale is averaged, so the position of the gap kind of loses meaning

Reply:
And: it's never late to join the party, thanks for reviewing! Lmk if you still think it's a problematic approach